docs/src/logging.md

```diff
@@ -0,0 +1,195 @@
+# [Logging](@id sec_logging)
+
+In the final part of the square‑root story we instrument the Heron iteration with logging.
+The logging system answers two questions separately:
+
+* **What happens when we log?** → a [`LoggingAction`](@ref).
+* **When do we log?** → an [`AlgorithmLogger`](@ref) mapping contexts to actions.
+
+This separation lets users compose rich behaviors (printing, collecting stats, plotting) without modifying algorithm code, and lets algorithm authors emit domain‑specific events.
+
+## Recap: Default logging contexts
+
+The generic `solve!` loop in `interface/interface.jl` emits these contexts:
+
+| Context    | Meaning                                  |
+|-----------:|------------------------------------------|
+| `:Start`   | Just before iteration begins             |
+| `:Init`    | After state initialization               |
+| `:PreStep` | Right before a `step!`                   |
+| `:PostStep`| Right after a `step!`                    |
+| `:Stop`    | After halting (stopping criterion met)   |
+
+Algorithms may add their own (e.g. `:Refinement`, `:RejectedStep`) by calling [`handle_message`](@ref) with a custom symbol.
+
+## Global enable/disable
+
+```julia
+using AlgorithmsInterface: set_global_logging_state!, get_global_logging_state
+set_global_logging_state!(false)  # silence all logging
+previous = set_global_logging_state!(true)  # restore
+```
+
+## Instrumenting the square‑root algorithm
+
+We start with minimal feedback: iteration number and current estimate.
+
+```julia
+iter_printer = CallbackAction() do alg, prob, st
+    @printf("Iter %3d  x = %.12f\n", st.iteration, st.x)
+end
+
+logger = AlgorithmLogger(:PostStep => iter_printer)
+
+with(ALGORITHM_LOGGER => logger) do
+    state = solve(SqrtProblem(42.0), HeronAlgorithm(1.0, StopAfterIteration(8)))
+    println("Final ≈ sqrt(42): ", state.x)
+end
+```
+
+### Timing the run
+
+Add actions for start timestamp and per‑step elapsed time:
+
+```julia
+start_time = Ref{Float64}(0.0)
+
+record_start = CallbackAction() do alg, prob, st
```

Comment:
a very short question about these examples: Usually we had `problem, algorithm, state` as order of arguments, why are these switched here?

Also in naming, within the package I now use full names, Here it is abbreviated, which one do we favour? (Or: Are we maybe ok with not doing that consistently and in long lines we shorten variables?)

Reply:
I still need to adapt this to be the correct order, I got a little confused and forgot what we settled on, and apparently the stopping criterion still weren't consistent. Did we in the end decide to always do problem algorithm state, even with mutating? I vaguely recall that, but a part of the docs still said otherwise 😅
I am okay with shortening things, I'll probably expand some of the names myself here, since I still haven't really internalized the common abbreviations, but I see how it can be annoying :)

Reply:
I think we decided on problem-algorithm-state as a consistent thing, and usually using full variable names - but we have the notations page, where we can then also introduce short variants we use in appropriate places.